File: backtrack/expression_add_operators/expression_add_operators.jl
function product(arr, repeat)
  arrs = [arr for _ in 1:repeat]
  Iterators.product(arrs...)
end

function solve(num, target)
    n = length(num)
    return Channel() do c
        for comb in product(["", "+", "*", "-"], n - 1)
            solution = []
            # interleave numbers with operators
            for i in 1:n - 1
                push!(solution, num[i])
                push!(solution, comb[i])
            end
            push!(solution, num[end])
            solution = join(solution, "")
            if eval(parse(solution)) == target
                push!(c, solution)
            end
        end
    end
end

# "123", 6 -> ["1+2+3", "1*2*3"]
println(collect(solve("123", 6)))

# "232", 8 -> ["2*3+2", "2+3*2"]
println(collect(solve("232", 8)))

# "123045", 3 -> ['1+2+3*0*4*5', '1+2-3*0*4*5', '1*2+3*0-4+5', '1*2-3*0-4+5', '1-2+3+0-4+5', '1-2+3-0-4+5']
println(collect(solve("123045", 3)))

println(collect(solve("105", 5)))

File: backtrack/expression_add_operators/expression_add_operators.rb
def product(arr, repeat)
  arrs = (1..repeat - 1).map { arr }
  arr.product(*arrs)
end

def solve(num, target)
  n = num.length
  Enumerator.new do |enum|
    product(['', '+', '*', '-'], n - 1).each do |comb|
      solution = []
      # interleave numbers with operators
      (0...n - 1).each do |i|
        solution << num[i]
        solution << comb[i]
      end
      solution << num[n - 1]
      solution = solution.join('')
      enum.yield solution if eval(solution) == target
    end
  end
end

# "123", 6 -> ["1+2+3", "1*2*3"]
p solve('123', 6).to_a

# "232", 8 -> ["2*3+2", "2+3*2"]
p solve('232', 8).to_a

# "123045", 3 -> ['1+2+3*0*4*5', '1+2-3*0*4*5', '1*2+3*0-4+5', '1*2-3*0-4+5', '1-2+3+0-4+5', '1-2+3-0-4+5']
p solve('123045', 3).to_a

p solve('105', 5).to_a
